components/eamxx/src/physics/dp/impl/dp_iop_setinitial_impl.hpp

```diff
@@ -12,11 +12,203 @@ namespace dp {
  */
 
 template<typename S, typename D>
-KOKKOS_FUNCTION
-void Functions<S,D>::iop_setinitial(const Int& nelemd, const uview_1d<element_t>& elem)
+void Functions<S,D>::iop_setinitial(
+    const Int& plev,
+    const Int& pcnst,
+    const Int& nelemd,
+    const Int& np,
+    const Int& nstep,
+    const bool& use_replay,
+    const bool& dynproc,
+    const bool& have_t,
+    const bool& have_q,
+    const bool& have_ps,
+    const bool& have_u,
+    const bool& have_v,
+    const bool& have_numliq,
+    const bool& have_cldliq,
+    const bool& have_numice,
+    const bool& have_cldice,
+    const bool& scm_zero_non_iop_tracers,
+    const bool& is_first_restart_step,
+    const uview_1d<const Scalar>& qmin,
+    const uview_1d<const Spack>& uobs,
+    const uview_1d<const Spack>& vobs,
+    const uview_1d<const Spack>& numliqobs,
+    const uview_1d<const Spack>& numiceobs,
+    const uview_1d<const Spack>& cldliqobs,
+    const uview_1d<const Spack>& cldiceobs,
+    const Scalar& psobs,
+    const uview_1d<const Scalar>& dx_short,
+    Scalar& dyn_dx_size,
+    tracer_t& tracers,
+    element_t& elem,
+    const uview_1d<Spack>& tobs,
+    const uview_1d<Spack>& qobs)
 {
-  // TODO
-  // Note, argument types may need tweaking. Generator is not always able to tell what needs to be packed
+  // Made these up
+
+  //FieldGroup g = ...;
+  //const auto& names = g.m_info->m_field_names;
+  constexpr Int inumliq = 1;
+  constexpr Int inumice = 2;
+  constexpr Int icldliq = 3;
+  constexpr Int icldice = 4;
+
+  const Int plev_packs = ekat::npack<Spack>(plev);
+  const auto policy = ekat::ExeSpaceUtils<ExeSpace>::get_default_team_policy(1, plev_packs);
+
+  assert(np <= NP);
+  //assert(plev_packs <= Homme::NUM_LEV);
+  assert(tracers.inited());
+  assert(elem.inited());
+
+  if (!use_replay && nstep == 0 && dynproc) {
+
+    Kokkos::parallel_for(
+      "iop_setinitial loop",
+      policy,
+      KOKKOS_LAMBDA(const MemberType& team) {
+
+      // We cannot support parallelism at the element level when nstep<=1 because the thelev
+      // computation of tobs is dependent on prior iterations that may have alterted tobs
+      for (Int ie = 0; ie < nelemd; ++ie) {
+        for (Int j = 0; j < np; ++j) {
+          for (Int i = 0; i < np; ++i) {
+
+          // Find level where tobs is no longer zero
+          Int thelev=-1;
+          if (nstep <= 1) {
```

Comment:
All these `if (nstep <= 1)` and `if (nstep == 0)` seem unnecessary since part of the if-statement these are contained in is `if (... && nstep == 0 && ...)`. I may be missing something.

Reply:
Nice catch. That will reduce some of mess inside.